File: frontends/Julia/Exasim/src/Gencode/cmakecompile_combined.jl
# Build ONE solver executable that runs several generated models together
# (combined multi-PDE). Each pde in `pdes` must already have its kernels
# generated (Gencode.gencode) into its own per-model build dir and a distinct
# modelnumber (hence distinct modelid). Renders the frontend-app-combined
# templates into <builddir>/combined and builds against the installed Exasim.
# Mirrors the Python frontend's cmakecompile_combined.
function cmakecompile_combined(pdes)
    print("Compile combined multi-PDE Exasim app against the installed Exasim package...\n")

    prefix = install_prefix()
    p0 = pdes[1]
    builddir = joinpath(p0.builddir, "combined")
    mkpath(builddir)

    ids = Int[]
    kdirs = String[]
    for p in pdes
        p.exasimpath = prefix
        push!(ids, resolve_modelid(p))
        kd = joinpath(model_builddir(p), "kernels")
        isdir(kd) || error("No generated kernels at $kd; run Gencode.gencode(pde) first.")
        push!(kdirs, kd)
    end
    length(unique(ids)) == length(ids) ||
        error("combined models need distinct modelids (one per slot); got $ids")

    if p0.platform == "gpu"
        variant = p0.mpiprocs > 1 ? "gpumpi" : "gpu"
    else
        variant = p0.mpiprocs > 1 ? "cpumpi" : "cpu"
    end

    tmpl = frontend_app_combined_template_dir()
    subs = Dict(
        "EXASIM_VARIANT" => variant,
        "MODEL_IDS" => join(string.(ids), ", "),          # main.cpp: {100, 101}
        "MODEL_ID_LIST" => join(string.(ids), " "),       # CMake: IDS 100 101
        "KERNEL_DIRS" => join(["\"" * d * "\"" for d in kdirs], " "),  # KERNELS_DIRS
    )
    render(joinpath(tmpl, "CMakeLists.txt.in"), joinpath(builddir, "CMakeLists.txt"), subs)
    render(joinpath(tmpl, "main.cpp.in"), joinpath(builddir, "main.cpp"), subs)

    bdir = joinpath(builddir, "build")
    exe = joinpath(bdir, "exasimapp")
    cmake = cmake_command()
    cfg = `$cmake -S $builddir -B $bdir -DExasim_DIR=$(cmake_dir())`
    run(cfg)
    jobs = get(ENV, "JOBS", string(Sys.CPU_THREADS))
    run(`$cmake --build $bdir --parallel $jobs`)
    isfile(exe) || error("Build did not produce $exe.")
    return string(cfg)
end

File: frontends/Matlab/Gencode/cmakecompile_combined.m
function comstr = cmakecompile_combined(pdes)
% Build ONE solver executable that runs several generated models together
% (combined multi-PDE). Each pde in the cell array `pdes` must already have its
% kernels generated (kkgencode) into its own per-model build dir and a distinct
% modelnumber (hence distinct modelid). Renders the frontend-app-combined
% templates into <builddir>/combined and builds against the installed Exasim.
% Mirrors the Python frontend's cmakecompile_combined.

disp("Compile combined multi-PDE Exasim app against the installed Exasim package...");

prefix = exasim_install_prefix();
cmakecmd = exasim_cmake_command(prefix);

p0 = pdes{1};
builddir = string(p0.builddir) + "/combined";
if ~exist(char(builddir), 'dir'), mkdir(char(builddir)); end

n = numel(pdes);
ids = strings(1, n);
kdirs = strings(1, n);
for m = 1:n
    pdes{m}.exasimpath = prefix;
    ids(m) = string(resolve_modelid(pdes{m}));
    kd = model_builddir(pdes{m}) + "/kernels";
    if ~exist(char(kd), 'dir')
        error("No generated kernels at %s; run kkgencode(pde) first.", kd);
    end
    kdirs(m) = """" + kd + """";
end
if numel(unique(ids)) ~= n
    error("combined models need distinct modelids (one per slot); got %s", strjoin(ids, " "));
end

if p0.platform == "gpu"
    if p0.mpiprocs > 1, variant = "gpumpi"; else, variant = "gpu"; end
else
    if p0.mpiprocs > 1, variant = "cpumpi"; else, variant = "cpu"; end
end

tmpl = prefix + "/lib/cmake/Exasim/frontend-app-combined";
subs = {"EXASIM_VARIANT", variant; ...
        "MODEL_IDS",     strjoin(ids, ", "); ...    % main.cpp: {100, 101}
        "MODEL_ID_LIST", strjoin(ids, " "); ...     % CMake: IDS 100 101
        "KERNEL_DIRS",   strjoin(kdirs, " ")};      % CMake: KERNELS_DIRS
rendertemplate(tmpl + "/CMakeLists.txt.in", builddir + "/CMakeLists.txt", subs);
rendertemplate(tmpl + "/main.cpp.in", builddir + "/main.cpp", subs);

bdir = builddir + "/build";
exe = bdir + "/exasimapp";
comstr = cmakecmd + " -S " + builddir + " -B " + bdir + ...
         " -DExasim_DIR=" + prefix + "/lib/cmake/Exasim";
runchecked(comstr);

jobs = getenv('JOBS');
if isempty(jobs), jobs = num2str(feature('numcores')); end
runchecked(cmakecmd + " --build " + bdir + " --parallel " + jobs);

if ~exist(char(exe), 'file')
    error("Build did not produce %s.", exe);
end
end

function rendertemplate(srcfile, dstfile, subs)
text = fileread(char(srcfile));
for i = 1:size(subs, 1)
    text = strrep(text, char("@" + subs{i,1} + "@"), char(string(subs{i,2})));
end
if exist(char(dstfile), 'file') == 2 && strcmp(fileread(char(dstfile)), text)
    return;
end
fid = fopen(char(dstfile), 'w');
fwrite(fid, text);
fclose(fid);
end

function runchecked(cmd)
[status, output] = system(char(cmd));
disp(output);
if status ~= 0
    error("Command failed (exit %d): %s", status, cmd);
end
end
